web/elm/src/Build/StepTree/StepTree.elm

```diff
@@ -1257,41 +1234,76 @@ viewStepHeaderLabel headerType changed stepID =
-                            (Dict.fromList
-                                [ ( "initialization"
-                                  , Html.text (Duration.format initDuration)
-                                  )
-                                , ( "step", Html.text (Duration.format stepDuration) )
-                                ]
-                            )
-                        ]
-                    ]
-                , Html.div
-                    Styles.durationTooltipArrow
-                    []
-                ]
+tooltip : StepTreeModel -> { a | hovered : HoverState.HoverState } -> Maybe Tooltip.Tooltip
+tooltip model { hovered } =
+    case hovered of
+        HoverState.Tooltip (ChangedStepLabel _ text) _ ->
+            Just
+                { body =
+                    Html.div
+                        Styles.changedStepTooltip
+                        [ Html.text text ]
+                , attachPosition =
+                    { direction = Tooltip.Top
+                    , alignment = Tooltip.Start
+                    }
+                , arrow = Just { size = 5, color = Colors.tooltipBackground }
+                }
 
-            _ ->
-                []
+        HoverState.Tooltip (StepHeaderImageFetching _) _ ->
+            Just
+                { body =
+                    Html.div
+                        Styles.changedStepTooltip
+                        [ Html.text "initialization" ]
+                , attachPosition =
+                    { direction = Tooltip.Top
+                    , alignment = Tooltip.End
+                    }
+                , arrow = Just { size = 5, color = Colors.tooltipBackground }
+                }
 
-    else
-        []
+        HoverState.Tooltip (StepState id) _ ->
+            Dict.get id model.steps
+                |> Maybe.map stepDurationTooltip
+
+        _ ->
+            Nothing
+
+
+stepDurationTooltip : Step -> Tooltip.Tooltip
+stepDurationTooltip { initialize, start, finish } =
+    { body =
+        Html.div Styles.durationTooltip
+            [ case ( initialize, start, finish ) of
+                ( Just initializedAt, Just startedAt, Just finishedAt ) ->
+                    let
+                        initDuration =
+                            Duration.between initializedAt startedAt
+
+                        stepDuration =
+                            Duration.between startedAt finishedAt
+                    in
+                    DictView.view []
+                        (Dict.fromList
+                            [ ( "initialization"
+                              , Html.text (Duration.format initDuration)
+                              )
+                            , ( "step"
+                              , Html.text (Duration.format stepDuration)
+                              )
+                            ]
+                        )
+
+                ( Nothing, Nothing, Nothing ) ->
+                    Html.text "pending"
```

Comment:
Really minor, but for steps that'll never run (e.g. hooks that aren't run, steps after a failed step in a `do`, implicit `get` after failed `put`), the `pending` tooltip is a bit misleading:

<img width="502" alt="Screen Shot 2020-10-18 at 10 24 03 PM" src="https://user-images.githubusercontent.com/26583442/96395027-8426d680-1191-11eb-8e29-e3257e0a6b42.png">

Reply:
Unrelated, and you may have noticed this, but the expanded icon doesn't look so great when the step header has a border:

<img width="490" alt="Screen Shot 2020-10-18 at 10 38 44 PM" src="https://user-images.githubusercontent.com/26583442/96395536-dddbd080-1192-11eb-857e-a42d1df68112.png">

Reply:
Dang, yeah that's because I changed the border to a `box-shadow` because I noticed the headers became 1px bigger on all sides, bringing them out of alignment with other elements on the resource page. But now the shadow is below that element.

I think I'll change it back to borders, since I also noticed some weirdness on HiDPI displays - it seemed like the upper border was ~0.5px thicker than the lower one, and I'm guessing that's just more weirdness with box-shadow being used this way.

Will patch up the tooltip too. I'll have it show the step state if the timing info isn't available.

Reply:
Fixed the border interaction...kind of:

![image](https://user-images.githubusercontent.com/1880/96473255-6abc7380-11ff-11eb-8bea-e353feec35f6.png)

The trade-off is that the background color isn't flush to the edges when the step has no border:

![image](https://user-images.githubusercontent.com/1880/96473356-8a539c00-11ff-11eb-9157-1fcdcf61e313.png)

...but I think I'm OK with this since it's at least consistent, and fixing it would involve CSS trickery that's dependent on the step's state. I don't want to get too lost in the weeds on this since we might adjust it more later anyway.